src/main/kotlin/org/rust/lang/core/completion/CompletionEngine.kt

```diff
@@ -72,6 +73,22 @@ object CompletionEngine {
             ?.mapNotNull { it.libTarget }
             ?.map { LookupElementBuilder.create(extCrate, it.normName).withIcon(extCrate.getIcon(0)) }
             ?.toTypedArray() ?: emptyArray()
+
+    fun completeMod(mod: RsModDeclItem): Array<out LookupElement> {
+        val directory = mod.containingMod.ownedDirectory ?: return EMPTY_ARRAY
+
+        val modFromRsFile = directory.files
+            .filter { it.name.endsWith(".rs") }
+            .map { it.name.substringBeforeLast('.') }
+            .filter { it != "mod" }
```

Comment:
Yes, Grammar Kit generate a parser based on the rules defined in the `.bnf` file. The parser, in its turn, produces PSI trees from the source code. [Here's the overview](https://github.com/intellij-rust/intellij-rust/blob/master/ARCHITECTURE.md) of how it works and links to the Grammar Kit documentation.